### DISCUSSION TITLE
Few questions to get started with Qstash through Upstash?

### DISCUSSION THREAD

**Discussion by lancejpollard** [upstash/product-roadmap#90] [Announcements]
The docs are very scarce for qstash so I wanted to clarify a few things before getting started (assuming you are still developing this).

For receiving a message:

```ts
import { Receiver } from "@upstash/qstash";

const r = new Receiver({
  currentSigningKey: "..",
  nextSigningKey: "..",
});

const isValid = await r.verify({
  /**
   * The signature from the `Upstash-Signature` header.
   *
   * Please note that on some platforms (e.g. Vercel or Netlify) you might
   * receive the header in lower case: `upstash-signature` 
   *
   */
  signature: "string";

  /**
   * The raw request body.
   */
  body: "string";

  /**
   * Number of seconds to tolerate when checking `nbf` and `exp` claims, to deal with small clock differences among different servers
   *
   * @default 0
   */
  clockTolerance?: number;
})
```

- What should the values of `currentSigningKey` and `nextSigningKey` be, where do those come from?
- When should I care about `clockTolerance` and when should I set it, what should I set it to?

When publishing a message:

```ts
const res = await c.publishJSON({
  url: "https://my-api...",
  // or topic: "the name or id of a topic"
  body: {
    hello: "world",
  },
});
console.log(res);
// { messageId: "msg_xxxxxxxxxxxxxxxx" }
```

- What should/can I do with `messageId` in the response?
- Is this a POST or GET request to my endpoint `url`?
- What should my endpoint respond with?
  - If there is an error for some reason in my processing of the job, what should I do (not a code bug, but a processing bug like there was an error in the file conversion format or something).
  - If there is an error otherwise, what should I respond with?
  - If there is a success, what should I respond with?

If you could show a quick `express` app example, that would be quite helpful, though I am using Next.js if that works too.

- What does pubishing to a topic do / how does it work? Does it hit my endpoint? Or how do I integrate it with my app?

Also

- What if my job takes 10 minutes to complete (like video processing), do I respond early or wait until job completion before responding? I am new to queues/jobs, so looking for some basic guidance.

Thank you for the help.

**Comment by sancar** [comment]
Hi @lancejpollard , the answers are inlined. 

> What should the values of currentSigningKey and nextSigningKey be, where do those come from?
They are available on the console. 
<img width="849" alt="Screenshot 2024-02-06 at 07 28 07" src="https://github.com/upstash/issues/assets/570462/fe81b146-1f8c-4ced-9570-7ba33cfae479">

And also you can reach them from rest API: 
https://upstash.com/docs/qstash/api/signingKeys/get

> When should I care about clockTolerance and when should I set it, what should I set it to?

~We don't have any config called `clockTolerance`. Where did you saw this ?~

A team mate pointed me the source https://github.com/upstash/sdk-qstash-ts. 

It is for advanced case and default value should be enough for almost all cases. I have updated the code in Readme and removed it to avoid confusion. You can check following doc for [Verify Signatures](https://upstash.com/docs/qstash/howto/signature)

> What should/can I do with messageId in the response?

You can use it to match with the event logs/dlqs later when needed. 
Another example, some users store the message id along with a context. And they use that context upon [callbacks](https://upstash.com/docs/qstash/features/callbacks) to take further action. This may not be needed for you, heavily depends on what you want to build.

Check out our API reference 
https://upstash.com/docs/qstash/api/events/list
https://upstash.com/docs/qstash/api/dlq/listMessages

> Is this a POST or GET request to my endpoint url?

The http `method` is configurable and `POST` by default. 
See all other parameters and their defaults here
https://upstash.com/docs/qstash/api/messages/create

> What should my endpoint respond with?

> If there is an error for some reason in my processing of the job, what should I do (not a code bug, but a processing bug like there was an error in the file conversion format or something).

 See 
 https://upstash.com/docs/qstash/features/retry
 `Many things can go wrong in a serverless environment. If your API does not respond with a success status code (2XX), we retry the request to ensure every message will be delivered.` 
             
 > If there is an error otherwise, what should I respond with?
 We don't distinquish these two, but if there is an error in the format and the message is doomed to fail every time it will eventually delivered to DLQ after retries are exhausted. I recommend using FailureCallbacks to handle dead messages. See https://upstash.com/docs/qstash/features/callbacks#what-is-a-failure-callback
             
 > If there is a success, what should I respond with?

Any 2xx code is fine.
             
> If you could show a quick express app example, that would be quite helpful, though I am using Next.js if that works too.

We have next js example here along with its code. 
https://upstash.com/examples/shoppingapp
Blog -> https://upstash.com/blog/session-store-with-clerk
Code -> https://github.com/upstash/examples/blob/main/examples/session-store-with-clerk/README.md
             
> What does pubishing to a topic do / how does it work? Does it hit my endpoint? Or how do I integrate it with my app?

See https://upstash.com/docs/qstash/features/topics
             
> What if my job takes 10 minutes to complete (like video processing), do I respond early or wait until job completion before responding? I am new to queues/jobs, so looking for some basic guidance.

If you wait until the process is complete then it is easier to decide what failed and what is successful.         
If you just trigger and return, you need to keep track of failures yourselves, because Qstash will consider this a success. 
(There was a manual acknowledgement request for this case but not implemented yet) 
Having said that there is a limit depending on your plan how long you can wait. See `Http Connection Timeout` in [pricing](https://upstash.com/pricing/qstash).
